src/bitmap-adapter.js

```diff
@@ -0,0 +1,143 @@
+const log = require('./util/log');
+const base64js = require('base64-js');
+
+/**
+ * Adapts Scratch 2.0 bitmaps for use in scratch 3.0
+ */
+class BitmapAdapter {
+    /**
+     * @param {?function} makeImage HTML image constructor. Tests can provide this.
+     * @param {?function} makeCanvas HTML canvas constructor. Tests can provide this.
+     */
+    constructor (makeImage, makeCanvas) {
+        this._makeImage = makeImage ? makeImage : () => new Image();
+        this._makeCanvas = makeCanvas ? makeCanvas : () => document.createElement('canvas');
+    }
+
+    // Returns a canvas with the resized image
+    resize (image, newWidth, newHeight) {
+        // Resize in a 2 step process, matching width first, then height, in order
+        // to preserve nearest-neighbor sampling.
+        const stretchWidthCanvas = this._makeCanvas();
+        stretchWidthCanvas.width = newWidth;
+        stretchWidthCanvas.height = image.height;
+        let context = stretchWidthCanvas.getContext('2d');
+        context.imageSmoothingEnabled = false;
+        context.drawImage(image, 0, 0, stretchWidthCanvas.width, stretchWidthCanvas.height);
+        const stretchHeightCanvas = this._makeCanvas();
+        stretchHeightCanvas.width = newWidth;
+        stretchHeightCanvas.height = newHeight;
+        context = stretchHeightCanvas.getContext('2d');
+        context.imageSmoothingEnabled = false;
+        context.drawImage(stretchWidthCanvas, 0, 0, stretchHeightCanvas.width, stretchHeightCanvas.height);
+        return stretchHeightCanvas;
+    }
+
+    /**
+     * Scratch 2.0 had resolution 1 and 2 bitmaps. All bitmaps in Scratch 3.0 are equivalent
+     * to resolution 2 bitmaps. Therefore, converting a resolution 1 bitmap means doubling
+     * it in width and height.
+     * @param {!string} dataURI Base 64 encoded image data of the bitmap
+     * @param {!function} callback Called with the updated dataURI if conversion succeeded,
+     *     or no arguments if the conversion failed.
+     */
+    convertResolution1Bitmap (dataURI, callback) {
+        const image = this._makeImage();
+        image.src = dataURI;
+        image.onload = () => {
+            callback(this.resize(image, image.width * 2, image.height * 2).toDataURL());
```

Comment:
This is an unconventional use of callbacks and might be confusing if anyone else tries to use this function. In NodeJS callback functions usually take two arguments, error and result, in that order, and then they usually check if error is non-falsey first and do things with the error, otherwise they move on to the result. I think this code and the corresponding call to convertResolution1Bitmap in the VM should be update  to use the 'error first' callback structure.